greenplumpython/table.py

```diff
@@ -35,28 +30,93 @@ def __getitem__(self, key):
 
-    def select(self, expr_list: list):
+    def select(self, target_list: Iterable) -> "Table":
         return Table(
-            f"SELECT {','.join([str(target) for target in key])} FROM {self.name}", parents=[self]
+            f"""
+                SELECT {','.join([str(target) for target in target_list])} 
+                FROM {self._name}
+            """,
+            parents=[self],
         )
 
-    def describe(self):
-        raise NotImplementedError()
+    def columns(self) -> "Table":
+        if any(self._parents):
+            raise NotImplementedError()
+        return Table(
+            f"""
+                SELECT * 
+                FROM information_schema.columns 
+                WHERE table_name = quote_ident('{self._name}')
+            """,
+            db=self._db,
+        )
+
+    def name(self) -> str:
+        return self._name
+
+    def _list_lineage(self) -> Iterable["Table"]:
+        lineage = [self]
+        tables_visited = set()
+        current = 0
+        while current < len(lineage):
+            for table in lineage[current].parents:
+                if table._name not in tables_visited:
+                    lineage.append(table)
+                    tables_visited.add(table._name)
+            current += 1
+        return lineage
+
+    def _build_full_query(self) -> str:
+        if not any(self._parents):
+            return self._query
+        lineage = self._list_lineage()
+        cte_list = []
+        for table in reversed(lineage):
+            if table._name != self._name:
+                cte_list.append(f"{table._name} AS ({table._query})")
+        return "WITH " + ",".join(cte_list) + self.query
+
+    def fetch(self, all: bool = True) -> Iterable:
+        """
+        Fetch rows of this table.
+        - if all is True, fetch all rows at once
+        - otherwise, open a CURSOR and FETCH one row at a time
+        """
+        if not all:
+            raise NotImplementedError()
+        return self._db.execute(self._build_full_query())
+
+    def save_as(
+        self, table_name: str, temp: bool = True, column_names: Iterable[str] = []
+    ) -> "Table":
+        self._db.execute(
+            f"""
+            CREATE {'TEMP' if temp else ''} TABLE {table_name} ({','.join(column_names)}) 
+            AS {self._build_full_query()}
+            """,
+            has_results=False,
+        )
+        return table(table_name, self._db)
+
+
+# table_name can be table/view name
+def table(name: str, db: db.Database) -> Table:
+    return Table(f"TABLE {name}", name=name, db=db)
 
-    def name(self):
-        raise NotImplementedError()
 
-    # If all = false, use cursor to fetch
-    def fetch(all: bool = true):
-        raise NotImplementedError()
+def values(rows: Iterable[Tuple], db: db.Database) -> Table:
+    return Table(
+        f"VALUES {','.join(['(' + ','.join(str(datum) for datum in row) + ')' for row in rows])}",
+        db=db,
```

Comment:
```
gpadmin=# with a as ( VALUES (1, 2),(2),(3) ) select * from a;
ERROR:  VALUES lists must all be the same length
LINE 1: with a as ( VALUES (1, 2),(2),(3) ) select * from a;
```

And type checking is necessary.
I can create some tracker stories for those.

Reply:
Agree.